### DISCUSSION TITLE
APKS generated from AAB runs on some arm64-v8a devices while not on others.

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#1115] [Q&A]
After having successfully completed the aab->apks->install->run on a couple of arm64-v8a devices, I proceeded with uploading the aab file to the play store. All went well. I set up an internal test track, invited myself and some colleagues to test it out. I installed the test version on my phone (the one I use everyday and never for testing).

The app would not run. I was horrified. First, I ran back to the Play Console and immediately deleted the invitees list, so that I can avoid some uncomfortable questions.

Ok. After 4 hours of effort, I narrowed down the problem to a Native ANE. If I don't use a Native ANE (including it in the manifest is fine, but calling the ANE functions causes the problem), the app runs ok. **Again, the apk version works fine.** It is only the aab version fails.

Also, this happens on some devices only. This, I find strange. I guess, I should have tested the aab on my everyday phone before uploading it to the Play Store. But, why should that matter?

@ajwfrost - request your help in this matter.

Thanks.

PS: Deleted a lot of initial panic reaction from this post, as it seems irrelevant now.

------------------
Following are the logcat outputs for both devices:

**logcat of the device where the Native ANE Works (Micromax, Android 5.0.0):**
`08-31 02:06:23.231 I/air.com.mediapro.UISampler( 5018): ----- Testing MproXtraAnd -----` **(---> Java ANE)**
`08-31 02:06:23.232 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.232 I/air.com.mediapro.UISampler( 5018): function aneVersion() = 1`
`08-31 02:06:23.232 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.232 I/air.com.mediapro.UISampler( 5018): function getTestString() = Hello World. This is your ANE talking.`
`08-31 02:06:23.232 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.233 I/air.com.mediapro.UISampler( 5018): function aid() = 3dad079d7bba8128`
`08-31 02:06:23.233 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.234 I/air.com.mediapro.UISampler( 5018): function getExternalFilesDir() = /storage/emulated/0/Android/data/air.com.mediapro.UISampler/files`
`08-31 02:06:23.234 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.234 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.235 I/air.com.mediapro.UISampler( 5018): ----- WinSampleXtra can only be tested on Windows OS -----` **(---> AS EXPECTED)**
`08-31 02:06:23.235 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): ----- Testing SampleXtra -----` **(---> Native ANE)**
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): function hasInitialized() = true`
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): function aneVersion() = 2`
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): function isSupported() = true`
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): function getTestString() = This is a test string from ANE!`
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): function getHelloWorld() = Hello World! This is your ANE talking!`
`08-31 02:06:23.237 I/air.com.mediapro.UISampler( 5018): `

**logcat of the device where the Native ANE does not work (Nokia 6.1, Android 10):**
`08-31 09:28:32.969  8841  8841 I air.com.mediapro.UISampler: ----- Testing MproXtraAnd -----` **(---> Java ANE)**
`08-31 09:28:32.969  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.971  8841  8841 I air.com.mediapro.UISampler: function aneVersion() = 1`
`08-31 09:28:32.971  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.972  8841  8841 I air.com.mediapro.UISampler: function getTestString() = Hello World. This is your ANE talking.`
`08-31 09:28:32.973  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.976  8841  8841 I air.com.mediapro.UISampler: function aid() = fdb6f3ba11a2a67d`
`08-31 09:28:32.976  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.980  8841  8841 I air.com.mediapro.UISampler: function getExternalFilesDir() = /storage/emulated/0/Android/data/air.com.mediapro.UISampler/files`
`08-31 09:28:32.980  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.980  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.980  8841  8841 I air.com.mediapro.UISampler: ----- WinSampleXtra can only be tested on Windows OS -----` **(---> AS EXPECTED)**
`08-31 09:28:32.980  8841  8841 I air.com.mediapro.UISampler: `
`08-31 09:28:32.982  8841  8841 I air.com.mediapro.UISampler: ArgumentError: Error #3500: The extension context does not have a method with the name hasInitialized.` **(---> First function of the Native extension that is called.)**
`08-31 09:28:32.983  8841  8841 I air.com.mediapro.UISampler: 	at flash.external::ExtensionContext/_call()`
`08-31 09:28:32.983  8841  8841 I air.com.mediapro.UISampler: 	at flash.external::ExtensionContext/call()`
`08-31 09:28:32.983  8841  8841 I air.com.mediapro.UISampler: 	at com.mediapro::SampleXtra/hasInitialized()`
`08-31 09:28:32.983  8841  8841 I air.com.mediapro.UISampler: 	at screens.code::MainScreen$/test_SampleXtra()[D:\projects\Harman_Test\src\screens\code\MainScreen.as:92]`
`08-31 09:28:32.983  8841  8841 I air.com.mediapro.UISampler: 	at screens.code::MainScreen$/loadANETests()[D:\projects\Harman_Test\src\screens\code\MainScreen.as:59]`

**Both have been installed from the same aab.**

___
Seems to work on Android 8 and below. Devices with Android 9 (the ones I have tested) have a problem. Also, on Android 8 and below, the apks size is about 8mb, and on Android 9+, it is about 20mb.

**Comment by GeneralVimes** [comment]
I had the same issue because of 3rd party service: https://github.com/airsdk/Adobe-Runtime-Support/discussions/1039

The difference is, I integrate such services as IAP, Analytics, Ads not directly, with ANEs, but via Enhance.co: I built apk or aab, and then the code for the 3rd party service gets injected inside it.

We have worked together with Enhance support, and as a result, they added an option called "enhance-injectable-sdk-androidx", which I should select when I integrate 3rd party services into .aab. Since then everything works well.

I wonder, what should be changed in packaging the ANEs the usual way: something in the packaging function, or the ANE itself?...

**Comment by manishpatel0404** [comment]
Here's what I saw in the Play Console's pre-launch report, that I believe has something to do with loading the ANE:

`StrictMode policy violation: android.os.strictmode.NonSdkApiUsedViolation: Ljava/io/FileDescriptor;->descriptor:I`
`	at android.os.StrictMode.lambda$static$1(StrictMode.java:407)`
`	at android.os.-$$Lambda$StrictMode$lu9ekkHJ2HMz0jd3F8K8MnhenxQ.accept(Unknown Source:2)`
`	at com.adobe.air.Entrypoints.EntryMainWrapper(Native Method)`
`	at com.adobe.air.Entrypoints.EntryMain(Entrypoints.java:143)`
`	at com.adobe.air.AndroidActivityWrapper.LaunchApplication(AndroidActivityWrapper.java:1221)`
`	at com.adobe.air.AndroidActivityWrapper.launchApplication(AndroidActivityWrapper.java:1507)`
`	at com.adobe.air.AndroidActivityWrapper.onSurfaceInitialized(AndroidActivityWrapper.java:1493)`
`	at com.adobe.air.AIRWindowSurfaceView.surfaceChanged(AIRWindowSurfaceView.java:802)`
`	at android.view.SurfaceView.updateSurface(SurfaceView.java:942)`
`	at android.view.SurfaceView$2.onPreDraw(SurfaceView.java:215)`
`	at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:1102)`
`	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3310)`
`	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2200)`
`	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9037)`
`	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:999)`
`	at android.view.Choreographer.doCallbacks(Choreographer.java:797)`
`	at android.view.Choreographer.doFrame(Choreographer.java:732)`
`	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:984)`
`	at android.os.Handler.handleCallback(Handler.java:883)`
`	at android.os.Handler.dispatchMessage(Handler.java:100)`
`	at android.os.Looper.loop(Looper.java:237)`
`	at android.app.ActivityThread.main(ActivityThread.java:8016)`
`	at java.lang.reflect.Method.invoke(Native Method)`
`	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)`
`	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)`

And yet another:

`StrictMode policy violation: android.os.strictmode.NonSdkApiUsedViolation: Landroid/graphics/Typeface;->sSystemFontMap:Ljava/util/Map;`
`	at android.os.StrictMode.lambda$static$1(StrictMode.java:407)`
`	at android.os.-$$Lambda$StrictMode$lu9ekkHJ2HMz0jd3F8K8MnhenxQ.accept(Unknown Source:2)`
`	at java.lang.Class.getDeclaredField(Native Method)`
`	at androidx.test.tools.crawler.platform.hybrid.TextViewHelper.getSystemFontMap(TextViewHelper.java:3)`
`	at androidx.test.tools.crawler.platform.hybrid.TextViewHelper.<clinit>(TextViewHelper.java:1)`
`	at androidx.test.tools.crawler.platform.hybrid.TextViewHelper.getTextOfTextViewOrEmpty(TextViewHelper.java:1)`
`	at androidx.test.tools.crawler.platform.hybrid.EspressoElement.extractDescription(EspressoElement.java:4)`
`	at androidx.test.tools.crawler.platform.hybrid.EspressoElement.create(EspressoElement.java:2)`
`	at androidx.test.tools.crawler.platform.hybrid.ViewScreenStateBuilder$ViewTreeNode.<init>(ViewScreenStateBuilder.java:1)`
`	at androidx.test.tools.crawler.platform.hybrid.ViewScreenStateBuilder.traverseAndBuild(ViewScreenStateBuilder.java:5)`
`	at androidx.test.tools.crawler.platform.hybrid.HybridStateExtractor.tryExtractingScreenState(HybridStateExtractor.java:9)`
`	at androidx.test.tools.crawler.platform.hybrid.HybridStateExtractor.getStableScreen(HybridStateExtractor.java:2)`
`	at androidx.test.tools.crawler.platform.RemotePlatform.handlePerformScrape(RemotePlatform.java:1)`
`	at androidx.test.tools.crawler.platform.RemotePlatform.messageLoop(RemotePlatform.java:27)`
`	at androidx.test.tools.crawler.platform.RemotePlatform.lambda$startCrawlAndWaitUntilFinished$0$RemotePlatform(RemotePlatform.java:1)`
`	at androidx.test.tools.crawler.platform.RemotePlatform$$ExternalSyntheticLambda0.run(Unknown Source:6)`
`	at java.lang.Thread.run(Thread.java:919)`

I don't understand either of these. If there is something I am doing wrong, please let me know.

Thanks.

**Reply by ajwfrost** [reply]
Hi
Those two reports from StrictMode are both for API usage that is in the "greylist" i.e. it's allowed, but liable to change in the future. So, they should not be what's causing the problem.
Could we get the AAB file plus the APKS files that you generated please, so we can inspect the difference? it's most likely caused by a problem in loading the native library and then finding the function within it (probably, the context initializer function, to get that error).
If you can build an AAB file with "aab-debug" type, it would help, as we added a little more logging in there plus it generates start-up code that means we can get it to load our own libCore.so (at least, assuming Android security hasn't blocked that mechanism yet..)
thanks
https://transfer.harman.com/requests/X0sLFbcKszxRaC5hGmftxD

**Reply by ajwfrost** [reply]
Also just noted from your original post (is this still the case?)
> So I unzipped both APKS files into 2 folders and observed:
> 8.14 mb for the base-arm64_v8a.apk that worked.
> 20.7 mb for the base-arm64_v8a.apk that did not work.
>Digging further, I unzipped both base-arm64_v8a.apk into 2 folders and observed the files inside were identical in size.

That might imply a problem writing the zip contents block... which could then mask another problem.. So definitely we should look at that part of things..

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - I have uploaded the files for your investigation. If you extract the `base-arm64_v8a.apk` files from both apks, you will see that they have different sizes. However, when you extract the contents of the `base-arm64_v8a.apk` files, they extracted files report the same size. Even a binary compare of the contents conclude that they are the same files (except the timestamps, of course, which is expected).

**Reply by ajwfrost** [reply]
Hi - thanks for that, interesting that in one case the libraries are just being added (stored) without compression, and in the other they are being compressed as well. Both the zip files check out at least from what I can see..
So everything looks like it should, but for some reason the native file isn't being opened or read properly. We can check what's happening under the hood and get back to you...
thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - I had just dozed off.

Look forward to your reply. Thanks.

**Reply by ajwfrost** [reply]
:-) well it's still not finding the library ...
```
09-01 15:13:34.002 13523 13523 D AdobeAIR: dlopen failed: library "/data/data/air.com.mediapro.UISampler/lib/libcom.mediapro.SampleXtra.so" not found
```
And so, we find a utility class that's trying to use an API to get the appropriate folder for the binary, and falling back to the above choice. We don't actually know where the .so file is - or whether it's even been extracted from the APK, given how they're doing things now. Interesting thread below if you really want the details!
https://stackoverflow.com/questions/56551036/native-libraries-not-found-in-applicationinfo-nativelibrarydir-when-building-app

So we can do a few checks and changes to try to ensure that we can load the library on actual Android 9+ devices..

**Comment by manishpatel0404** [comment]
Based on the selected answer of the article, 

> By default, APKs generated from the Android App Bundle have the native libraries uncompressed on devices with Android M / API level 23 or higher...

So, is this what is happening? The `base-arm64_v8a.apk` is being retained in its uncompressed form, and not being extracted to the `lib`? And therefore, not able to open `libcom.mediapro.SampleXtra.so` from the `lib` folder because it is empty?

So, would you have to either add the flag `android.bundle.enableUncompressedNativeLibs=false` or let the Android platform load the library for you from the ucompressed apk file, using `System.loadLibrary`?

Also, I read somewhere that"

> WARNING: The option setting 'android.bundle.enableUncompressedNativeLibs=false' is experimental and unsupported.

Which means, going forward, they essentially want you to use the Android platform's `System.loadLibrary`?

Or maybe use, `android_dlopen_ext` instead of `dlopen`?

Did I understand this correctly?

Thanks.

**Reply by manishpatel0404** [reply]
Also, this probably answers your point:
>interesting that in one case the libraries are just being added (stored) without compression, and in the other they are being compressed as well.

Just thinking aloud, upto Android 8, the bundle installation process may be internally following the old apk install approach. From Android 9 onwards, the `base-arm64_v8a.apk` is stored in its uncompressed form, even though ADT might have compressed it. Maybe this is why the `base-arm64_v8a.apk` is not compressed.

**Reply by ajwfrost** [reply]
@manishpatel0404 yes exactly, a fair number of different options there for how to do it, but of course, no one solution that will work on Android SDK 14 onwards.. but it's fairly simple to take a cut-off and use one of the simpler mechanisms for the more recent handsets.

So can you please try using the "runtimeClasses.jar" from the below link, which should then mean that the C++ code is able to load the library properly now. One of the guys here was able to reproduce your issue (when using your ANE and a simple test app) on an Android 9 handset, and this has fixed it..
https://transfer.harman.com/message/rEop5q61T3HohYKA17ke2F

thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - YEP. It works. I will now try it on the Play Store and report when I am done. I am confident it will work on the Play Store as well, because the root cause seems to have been sorted out.

Is it possible to share which option you finally ended up taking?

Also, how come this did not come up earlier. I am assuming that not many developers use Native Code ANEs. Is this assumption correct? If so, it is sad. There is so much one can do with Native Code ANEs. I will open up a different discussion post on this and see what the community has to share about this.

Thanks, once again. You guys are great.

**Reply by ajwfrost** [reply]
Great, thanks. Actually we're just going with the "use the .so filename with no path, and trust the OS to work out where to load it from" option :-)
You'd hope this would work perhaps, since we're loading it from the same path as the library that loads it (libCore.so i.e. the core AIR runtime library...)

I agree re. the native code, I think in Android most of the ANEs just use Java which to be fair is the main language that Android developers use (or used, I've not really investigated Kotlin). But for efficiency and power then yes I personally prefer C++. I'd actually be more inclined though - unless there are specific API/dependencies - to use Crossbridge and compile the C++ into a SWF and avoid ANEs altogether!

cheers

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - I signed out of all accounts, cleaned out the browser cache and uploaded the aab file. It got uploaded without errors this time and I can now install the app from the internal test track. It works as expected on Android 10, as well, just like the offline install. I have a few colleagues who have Android 9 and 11. I will add them to the test track so that I can get a complete feedback.

If you wish, I can add a couple of test email ids, so that you can also have a look.

Thanks, once again.

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - In my haste, I didn't see your comment about Crossbridge.

We have been using Flash for creating our K-12 animation and simple interactivity for many years, I was more Gung ho on Director (another excellent product, neither Macromedia, nor Adobe could sustain). We used the FlashAsset Xtra to access the swfs embedded into Director movies and created all front-ends using Director.

I had never heard of Crossbridge and having briefly gone through its documentation on github where Adobe open-sourced it, it seems very very promising. I will try it out when I can spare some more time and see how one of our C-based ANEs can be converted to swf, using Crossbridge. In fact, it could just be another feather is AIR's cap, to be able to provide an easier route to including native code in AIR apps, over and above the ANE route.

**Reply by marchbold** [reply]
Crossbridge is great and a much better approach for pure c/c++ libs, simpler integration once you have the lib building and definitely worth checking it out!

**Reply by manishpatel0404** [reply]
It certainly looks good @marchbold. Will definitely try it out. I have also seen the great effort you are putting into [https://airsdk.dev/](url). I would be glad to contribute to this effort, once I am done porting our Director applications to AIR. One of them is already on the Play Console's internal testing track, thanks to the guidance of @ajwfrost and team.

**Reply by ajwfrost** [reply]
@manishpatel0404 can I just check - having seen a notification email from last night - did you get it uploaded to the Play Console properly? Did it only work after the 'sign out and cleaned the cache' bit?
thanks

**Reply by manishpatel0404** [reply]
Hi @ajwfrost - Yes, I signed out of Play Console, cleaned up the browser cache and signed back in, after which the upload went through without error. Since the upload had failed a few times before, I thought that Play Console might have stored some session data locally, which might be causing the issue. However, I am not sure it this was the case.

**Reply by manishpatel0404** [reply]
Just received a pre-launch report from Play Console. Other than the 2 "greylist", that had also appeared in the earlier upload, and a couple of accessibility issues about "Content labeling" and "Low contrast" (which I prefer to ignore, for the moment), everything else (Performance, Security and Trust) have no issues).

**Reply by ajwfrost** [reply]
Great thanks @manishpatel0404 for confirming!
